Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+
+        // This switch checks that the context element is supported and applies the
+        // same restrictions regarding content as the fast-path parser does for a
+        // corresponding nested tag.
+        // This is to ensure that we preserve correct HTML structure with respect
+        // to the context tag.
+        //
+        // If this switch has duplicate cases, then `tagNameHash()` needs to be
+        // updated.
+        switch (tagNameHash(contextTag.localName())) {
+#define TAG_CASE(TagName, TagClassName)                                                                      \
+        case tagNameHash(TagInfo::TagClassName::tagName):                                                    \
+            ASSERT(HTMLNames::TagName##Tag->localName().string().ascii() == TagInfo::TagClassName::tagName); \
+            if constexpr (!TagInfo::TagClassName::isVoid) {                                                  \
+                /* The hash function won't return collisions for the supported tags, but this function */    \
+                /* takes potentially unsupported tags, which may collide. Protect against that by */         \
+                /* checking equality. */                                                                     \
+                if (contextTag == HTMLNames::TagName##Tag) {                                                 \
+                    parseCompleteInput<typename TagInfo::TagClassName>();                                    \
+                    return !m_parsingFailed;                                                                 \
+                }                                                                                            \
+            }                                                                                                \
+        break;
+        FOR_EACH_SUPPORTED_TAG(TAG_CASE)
+        default:
+            break;
+#undef TAG_CASE
+        }
+
+        didFail(HTMLFastPathResult::FailedUnsupportedContextTag);
+        return false;
+    }
+
+    HTMLFastPathResult parseResult() const { return m_parseResult; }
+
+private:
+    CharSpan m_source;
+    Document& m_document;
+    DocumentFragment& m_fragment;
+
+    const Char* const m_end { m_source.data() + m_source.size() };
+    const Char* m_position { m_source.data() };
+
+    bool m_parsingFailed { false };
+    bool m_insideOfTagA { false };
+    // Used to limit how deep a hierarchy can be created. Also note that
+    // HTMLConstructionSite ends up flattening when this depth is reached.
+    unsigned m_elementDepth { 0 };
+    // 32 matches that used by HTMLToken::Attribute.
+    Vector<Char, 32> m_charBuffer;
+    Vector<UChar> m_ucharBuffer;
+    // Used if the attribute name contains upper case ascii (which must be mapped to lower case).
+    // 32 matches that used by HTMLToken::Attribute.
+    Vector<Char, 32> m_attributeNameBuffer;
+    Vector<Attribute> m_attributeBuffer;
```

Comment:
FYI, this vector is a big contributor to fastMalloc() in the profiles. I think ideally we'd use `Vector<Attribute, 10>` like we do in the full HTML parser. However, this requires some code changes to make it build/work. I am currently A/B testing such change.